adapta/storage/models/filter_expression.py

```diff
@@ -0,0 +1,204 @@
+"""
+    Models for generating filter expressions for PyArrow and Astra.
+"""
+
+from enum import Enum
+from abc import abstractmethod, ABC
+from typing import final, List, Dict, Generic, TypeVar, Any, Union, Type
+
+import pyarrow.compute
+from pyarrow.dataset import field as pyarrow_field
+
+TField = TypeVar("TField")  # pylint: disable=invalid-name
+TCompileResult = TypeVar("TCompileResult")  # pylint: disable=invalid-name
+
+
+class FilterExpressionOperation(Enum):
+    """
+    An enumeration of filter expression operations.
+    """
+
+    AND = {
+        "arrow": pyarrow.compute.Expression.__and__,
+        "astra": lambda left_exprs, right_exprs: [
+            left_expr | right_expr for left_expr in left_exprs for right_expr in right_exprs
+        ],
+    }
+    OR = {"arrow": pyarrow.compute.Expression.__or__, "astra": lambda left_exprs, right_exprs: left_exprs + right_exprs}
+    GT = {"arrow": pyarrow.compute.Expression.__gt__, "astra": "__gt"}
+    GE = {"arrow": pyarrow.compute.Expression.__ge__, "astra": "__gte"}
+    LT = {"arrow": pyarrow.compute.Expression.__lt__, "astra": "__lt"}
+    LE = {"arrow": pyarrow.compute.Expression.__le__, "astra": "__lte"}
+    EQ = {"arrow": pyarrow.compute.Expression.__eq__, "astra": ""}
+    IN = {"arrow": pyarrow.compute.Expression.isin, "astra": "__in"}
+
+
+@final
+class FilterField(Generic[TField]):
```

Comment:
I am not a fan of this required generic type. From a usability point of view it is quite verbose that we need to specify the type of the field every time we need to reference it. It looks like that Python 3.12 enables us to provide defaults. Until then, I think we should use Any type instead and do not enforce the generic typing.

https://peps.python.org/pep-0696/